contentcuration/kolibri_public/management/commands/rectify_incorrect_contentnode_source_fields.py

```diff
@@ -0,0 +1,157 @@
+            coalesced_license_id=Coalesce("license_id", -1),
+        )
+        diff_combined = diff.annotate(
+            original_source_node_f_changed=Exists(
+                original_source_nodes.filter(
+                    ~Q(coalesced_provider=OuterRef("coalesced_provider"))
+                    | ~Q(coalesced_author=OuterRef("coalesced_author"))
+                    | ~Q(coalesced_aggregator=OuterRef("coalesced_aggregator"))
+                    | ~Q(coalesced_license_id=OuterRef("coalesced_license_id"))
+                )
+            )
+        ).filter(original_source_node_f_changed=True)
+
+        final_nodes = diff_combined.values(
+            "id",
+            "channel_id",
+            "original_channel_id",
+            "original_source_node_id",
+            "coalesced_provider",
+            "coalesced_author",
+            "coalesced_aggregator",
+            "coalesced_license_id",
+            "original_source_node_f_changed",
+        ).order_by()
+
+        for item in final_nodes:
+            base_node = ContentNode.objects.get(pk=item["id"])
+
+            original_source_channel_id = item["original_channel_id"]
+            original_source_node_id = item["original_source_node_id"]
+            tree_id = (
+                Channel.objects.filter(pk=original_source_channel_id)
+                .values_list("main_tree__tree_id", flat=True)
+                .get()
+            )
+            original_source_node = ContentNode.objects.filter(
+                tree_id=tree_id, node_id=original_source_node_id
+            )
+
+            base_channel = Channel.objects.get(pk=item['channel_id'])
+
+            to_be_republished = not (base_channel.main_tree.get_family().filter(changed=True).exists())
+
+            if original_source_channel_id is not None and original_source_node.exists():
+                # original source node exists and its source fields dont match
+                # update the base node
+                if base_node.author != original_source_node[0].author:
+                    base_node.author = original_source_node[0].author
+                if base_node.provider != original_source_node[0].provider:
+                    base_node.provider = original_source_node[0].provider
+                if base_node.aggregator != original_source_node[0].aggregator:
+                    base_node.aggregator = original_source_node[0].aggregator
+                if base_node.license != original_source_node[0].license:
+                    base_node.license = original_source_node[0].license
+                base_node.save()
+                if to_be_republished and base_channel.last_published is not None:
+                    # we would repbulish the channel
+                    # Adding for testing
+                    if is_test:
+                        publish_channel(user_id, base_channel.id)
+                    else:
+                        publish_channel("SOME ID", base_channel.id, base_channel.id)
```

Comment:
My only uncertainty here is whether we should be running the publish as the user whose channel it is, or as some administrator.

The administrator means that we can always reliably run the publish as the same user, but does mean that it now appears that someone else has published the channel. The only other thing that comes to mind is that if we ran as the channel editor, they would receive an email indicating to them that their channel had been republished.